File: Sources/DesktopVoiceInput/Services/LLMClient.swift
import Foundation
import os

actor LLMClient {
    private static let logger = Logger(subsystem: "com.end.DesktopVoiceInput", category: "LLMClient")
    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    func complete(system: String, user: String, config: LLMProviderConfig) async throws -> String {
        let trimmed = config.trimmed
        switch trimmed.protocolType {
        case .openAICompatible:
            return try await openAIComplete(system: system, user: user, config: trimmed)
        case .anthropic:
            return try await anthropicComplete(system: system, user: user, config: trimmed)
        }
    }

    private func openAIComplete(system: String, user: String, config: LLMProviderConfig) async throws -> String {
        let endpoint = config.endpoint.hasSuffix("/")
            ? config.endpoint + "v1/chat/completions"
            : config.endpoint + "/v1/chat/completions"

        guard let url = URL(string: endpoint) else {
            throw LLMError.invalidEndpoint(config.endpoint)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("Bearer \(config.apiKey)", forHTTPHeaderField: "Authorization")

        let body: [String: Any] = [
            "model": config.model,
            "messages": [
                ["role": "system", "content": system],
                ["role": "user", "content": user]
            ],
            "temperature": 0.3
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw LLMError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            let errorBody = String(data: data, encoding: .utf8) ?? ""
            Self.logger.error("OpenAI API error. status=\(httpResponse.statusCode) body=\(errorBody, privacy: .public)")
            throw LLMError.apiError(statusCode: httpResponse.statusCode, message: extractErrorMessage(from: data) ?? "HTTP \(httpResponse.statusCode)")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let choices = json["choices"] as? [[String: Any]],
              let firstChoice = choices.first,
              let message = firstChoice["message"] as? [String: Any],
              let content = message["content"] as? String else {
            throw LLMError.invalidResponse
        }

        return content.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func anthropicComplete(system: String, user: String, config: LLMProviderConfig) async throws -> String {
        let endpoint = config.endpoint.hasSuffix("/")
            ? config.endpoint + "v1/messages"
            : config.endpoint + "/v1/messages"

        guard let url = URL(string: endpoint) else {
            throw LLMError.invalidEndpoint(config.endpoint)
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue(config.apiKey, forHTTPHeaderField: "x-api-key")
        request.addValue("2023-06-01", forHTTPHeaderField: "anthropic-version")

        let body: [String: Any] = [
            "model": config.model,
            "max_tokens": 4096,
            "system": system,
            "messages": [
                ["role": "user", "content": user]
            ]
        ]

        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw LLMError.invalidResponse
        }

        guard httpResponse.statusCode == 200 else {
            let errorBody = String(data: data, encoding: .utf8) ?? ""
            Self.logger.error("Anthropic API error. status=\(httpResponse.statusCode) body=\(errorBody, privacy: .public)")
            throw LLMError.apiError(statusCode: httpResponse.statusCode, message: extractErrorMessage(from: data) ?? "HTTP \(httpResponse.statusCode)")
        }

        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let content = json["content"] as? [[String: Any]],
              let firstBlock = content.first,
              let text = firstBlock["text"] as? String else {
            throw LLMError.invalidResponse
        }

        return text.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private func extractErrorMessage(from data: Data) -> String? {
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else { return nil }
        if let error = json["error"] as? [String: Any], let message = error["message"] as? String {
            return message
        }
        return json["message"] as? String ?? json["error"] as? String
    }
}

enum LLMError: LocalizedError {
    case invalidEndpoint(String)
    case invalidResponse
    case apiError(statusCode: Int, message: String)

    var errorDescription: String? {
        switch self {
        case .invalidEndpoint(let url): "LLM 端点地址无效：\(url)"
        case .invalidResponse: "LLM 返回了无法解析的响应"
        case .apiError(_, let message): "LLM 请求失败：\(message)"
        }
    }
}

File: windows/src/GuGuTalk.Core/Services/LLMClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using GuGuTalk.Core.Models;
using Serilog;

namespace GuGuTalk.Core.Services;

public sealed class LLMClient
{
    private static readonly ILogger Logger = Log.ForContext<LLMClient>();
    private readonly HttpClient _httpClient;

    public LLMClient()
    {
        _httpClient = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
    }

    public async Task<string> CompleteAsync(string system, string user, LLMProviderConfig config)
    {
        if (config.Endpoint.Contains("anthropic", StringComparison.OrdinalIgnoreCase))
            return await AnthropicCompleteAsync(system, user, config);

        return await OpenAICompleteAsync(system, user, config);
    }

    private async Task<string> OpenAICompleteAsync(string system, string user, LLMProviderConfig config)
    {
        string endpoint = config.Endpoint.TrimEnd('/') + "/v1/chat/completions";

        using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Headers.Add("Authorization", $"Bearer {config.ApiKey}");

        var body = new
        {
            model = config.Model,
            messages = new object[]
            {
                new { role = "system", content = system },
                new { role = "user", content = user }
            },
            temperature = 0.3
        };

        request.Content = JsonContent.Create(body);
        var response = await _httpClient.SendAsync(request);

        if (!response.IsSuccessStatusCode)
        {
            string errorBody = await response.Content.ReadAsStringAsync();
            Logger.Error("OpenAI API error. status={Status} body={Body}", (int)response.StatusCode, errorBody);
            throw new InvalidOperationException($"LLM 请求失败：HTTP {(int)response.StatusCode}");
        }

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        var root = doc.RootElement;
        var choices = root.GetProperty("choices");
        var content = choices[0].GetProperty("message").GetProperty("content").GetString();
        return content?.Trim() ?? "";
    }

    private async Task<string> AnthropicCompleteAsync(string system, string user, LLMProviderConfig config)
    {
        string endpoint = config.Endpoint.TrimEnd('/') + "/v1/messages";

        using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Headers.Add("x-api-key", config.ApiKey);
        request.Headers.Add("anthropic-version", "2023-06-01");

        var body = new
        {
            model = config.Model,
            max_tokens = 4096,
            system,
            messages = new object[]
            {
                new { role = "user", content = user }
            }
        };

        request.Content = JsonContent.Create(body);
        var response = await _httpClient.SendAsync(request);

        if (!response.IsSuccessStatusCode)
        {
            string errorBody = await response.Content.ReadAsStringAsync();
            Logger.Error("Anthropic API error. status={Status} body={Body}", (int)response.StatusCode, errorBody);
            throw new InvalidOperationException($"LLM 请求失败：HTTP {(int)response.StatusCode}");
        }

        using var doc = await JsonDocument.ParseAsync(await response.Content.ReadAsStreamAsync());
        var root = doc.RootElement;
        var content = root.GetProperty("content");
        var text = content[0].GetProperty("text").GetString();
        return text?.Trim() ?? "";
    }
}
